neevaco/neeva-android
# Make sure BillingFlow is safely launched in PLANS

Ensures the BillingFlow (which is launched by `subscriptionManager.buy(activity, tag)`) is memory-leak safe. 

Changes BillingFlow to be activated only on the `Plans` Screen. 

Also:
* Adds FirstRun simulator in Developer Settings
* Consolidate isPremiumPurchaseAvailable() into SubscriptionManager
* Moves select subscription tag state to SubscriptionManager + Shift where buy happens
* Restricts `buy()` to only happen in one place (`WelcomeFlowActivity`)
* Harnesses the callback from `onPurchasesUpdated()` to figure out if the user cancelled the billing flow or made a real purchase.

## Checklists

### Pointers to help with review
* https://www.figma.com/file/dylvDPB69JG9U1rzZ3HQ52/Android-Premium?node-id=0%3A1

### How was this tested?

### Manual test cases
Path: Launch BillingFlow from Settings ProfileRow + default browser is Neeva 
Expect: -> Returns to Settings and launches billing flow on top

Path: Launch BillingFlow from Settings ProfileRow + default browser is NOT Neeva 
Expect: -> Returns to WelcomeFlow SetDefaultBrowserScreen and launches billing flow on top

Path: Launch BillingFlow from Spaces Sign In + default browser is Neeva
Expect: -> Returns to Browser and launches billing flow on top

Path: Launch BillingFlow from Spaces Sign In + default browser is NOT Neeva 
Expect: -> Returns to WelcomeFlow SetDefaultBrowserScreen and launches billing flow on top

Path: Launch BillingFlow from ProfilePane + default browser is Neeva 
Expect: -> Returns to WelcomeFlow Settings and launches billing flow on top (Undefined behavior on the mocks)

Path: Launch BillingFlow from ProfilePane + default browser is NOT Neeva 
Expect: -> Returns to WelcomeFlow SetDefaultBrowser + launches billing flow on top (Undefined behavior on the mocks)

### Screenshots and screen recordings
Path: Launch BillingFlow from Settings ProfileRow + default browser is Neeva 
Expect: -> launches billing flow on top of PLANS and final destination is Settings.




https://user-images.githubusercontent.com/20916043/212532092-1406caf6-a0a2-47d9-a8cf-160e0173056b.mp4





Path: Launch BillingFlow from Settings ProfileRow + default browser is NOT Neeva 
Expect: -> launches billing flow on top of PLANS + shows set default browser + final destination is Settings.


https://user-images.githubusercontent.com/20916043/212532049-fb70a35a-12dc-485f-81f3-02507767027f.mp4





Path: Launch BillingFlow from Spaces Sign In + default browser is Neeva
Expect: -> Shows Plans + launches billing flow on top + Returns to Browser 


https://user-images.githubusercontent.com/20916043/212532146-d4fdc8ed-20ae-4783-b11f-487d93ab83ef.mp4



Path: Launch BillingFlow from Spaces Sign In + default browser is NOT Neeva 
Expect: -> launches billing flow on top of PLANS + shows set default browser + final destination is Browser.



https://user-images.githubusercontent.com/20916043/212532307-e0c42187-3def-4aac-9d2e-1760cabb2dc4.mp4





Path: Launch BillingFlow from ProfilePane + default browser is Neeva 
Expect: -> Shows Plans and launches billing flow on top + returns to Settings



https://user-images.githubusercontent.com/20916043/212532428-a9aa4ac6-f9e3-4732-90ff-a4ba0917f9e9.mp4




Path: Launch BillingFlow from ProfilePane + default browser is NOT Neeva 
Expect: -> Shows Plans and launches billing flow on top + shows set default browser screen + returns to Settings


https://user-images.githubusercontent.com/20916043/212532388-bb017be8-2c67-4934-82da-8ac17886ef4d.mp4
